src/ILCompiler.Compiler/src/IL/ILImporter.Scanner.cs

```diff
@@ -406,6 +406,12 @@ private void ImportCall(ILOpcode opcode, int token)
 
             TypeDesc exactType = method.OwningType;
 
+            if ((method.Signature.IsStatic && opcode == ILOpcode.callvirt)
+                || (method.IsNativeCallable && (opcode != ILOpcode.ldftn && opcode != ILOpcode.ldvirtftn)))
+            {
+                ThrowHelper.ThrowBadImageFormatException();
```

Comment:
Is there a way to attach more descriptive error message here?

This point of this change is make it easier to self-diagnose the problem. Today, I will get a crash at runtime and can get a clue that it has something to do with NativeCallable from the callstack. With this change, I will just get an non-descriptive bad image format exception and have to debug the compiler to figure out what is going on - it does not look like an improvement.

Reply:
We can throw a regular `Exception` and pass in the message or better still make ILC throw the exception at compile time, cos now I have to run the native binary to get the exception

Reply:
Yes, we just need a new `ExceptionStringID`, add a mapping to it in TypeLoaderExceptionHelper.cs, and add a new resource string in System.Private.CoreLib that the mapping will refer to.

I suggest using the ThrowHelper.ThrowInvalidProgramException(ExceptionStringID, MethodDesc) overload to avoid having to also pipe this through the IL emitter that emits throwing method bodies.